### DISCUSSION TITLE
virtual workspaces?

### DISCUSSION THREAD

**Discussion by thomase00** [Xpra-org/xpra#4055] [Ideas]
I haven't used xpra in a while, but one particular lacking feature that has kept me on VNC instead is support for virtual workspaces. I remember using Exceed's X Server WAY back and it supported this feature, even in seamless mode. The idea is simply to have the ability to assign individual windows to virtual workspaces (or pin a window to all workspaces), such that only windows from 1 virtual workspace are displayed at a given time. The system tray menu would have a way to select the current virtual workspace. The key is that this needs to work in seamless mode and remain persistent after disconnect and reconnect.

A feature like this is VERY important for users like me who maintain a persistent X session effectively indefinitely (i.e. until it crashes due to bugs, memory leaks, etc.) and need separate workspaces to manage windows related to different projects.

FWIW, my client machine runs Microsoft Windows 10, which also has its own virtual desktop/workspace feature. However, if I associate xpra windows with different Windows 10 virtual desktops, these associations are understandably lost after xpra session disconnect as they have nothing to do with xpra.

**Comment by totaam** [comment]
The window's mapping to workspaces should work with Linux X11 clients.
Wayland doesn't seem to have this exposed, at least not in GTK.

For MS Windows, I have found: https://github.com/mrob95/pyvda - it looks like it should be trivial to add a new backend.
I'm in the process of upgrading some hardware, I should have a functional MS Windows desktop to test with once that's complete.

**Comment by thomase00** [comment]
This seems like its tied to virtual desktop support on the client. Assuming it is possible to maintain state associating windows with client-side virtual desktops/workspaces that can be restored after suspend/reconnect, what happens if the reconnect is from a different client that doesn't have the same number of virtual desktops or doesn't support them at all? 

Wouldn't it be easier to handle this entirely on the server side, allowing the client to switch between different server-side workspaces that persist regardless of whether or not the client has native virtual desktop/workspace support?

**Reply by totaam** [reply]
> This seems like its tied to virtual desktop support on the client.

How could it be any different?

> what happens if the reconnect is from a different client

Each client has its own workspace mapping.

**Comment by totaam** [comment]
This is now implemented in https://github.com/Xpra-org/xpra/commit/1db7fbad2bcd9948bc3b69e2c3869f748120e958
You can find beta builds that implement this feature here: https://xpra.org/beta/windows/
(r34778 or later)

The code is really ugly and should be tidied up (moved to the `platform` modules), but this will do for now.
Can we close this?

**Reply by totaam** [reply]
Not heard back.
